File: 09-08-heap-y-cola-prio/adts/heap/min_heap.cpp
#pragma once
#include "heap.cpp"
#include <cassert>

template <class T> class minHeap : public heap<T> {
private:
  T *arr;
  int arrLength;
  int count;

  void grow(int newlength) {
    T *newArr = new T[newlength];
    for (int i = 0; i < this->arrLength; i++) {
      newArr[i] = this->arr[i];
    }

    //  delete[] this->arr;
    this->arr = newArr;
  }

  void swap(int posA, int posB) {
    T aux = this->arr[posA];
    this->arr[posA] = this->arr[posB];
    this->arr[posB] = aux;
  }

  void siftUp(int pos) {
    if (pos == 1) {
      return;
    }

    // comparo con el padre
    int parentPos = pos / 2;

    T elem = this->arr[pos];
    T parent = this->arr[parentPos];

    // si el padre es mayor ?
    if (parent > elem) {
      swap(pos, parentPos);
      siftUp(parentPos);
    }
  }

public:
  minHeap(int expectedSize) {
    this->arr = new T[expectedSize + 1];
    this->arrLength = expectedSize + 1;
    this->count = 0;
  }

  virtual void push(T elem) override {
    if (this->count == this->arrLength) {
      grow(this->arrLength * 2);
    }

    this->count++;
    this->arr[this->count] = elem;

    siftUp(this->count);
  }
  virtual T top() override {
    assert(size() >= 1);

    return this->arr[1];
  }
  virtual T pop() override { assert(false); }
  virtual int size() override { return this->count; }

  virtual bool isEmpty() override { return size() == 0; }
};

File: 09-08-heap-y-cola-prio/adts/heap/MinHeap.java
package adts.heap;

public class MinHeap<T extends Comparable<T>> implements Heap<T> {
   private Object[] arr;
   private int count;

   private void grow(int newlength) {
      Object[] newArr = new Object[newlength];
      for (int i = 0; i < this.arr.length; i++) {
         newArr[i] = this.arr[i];
      }

      this.arr = newArr;
   }

   private void swap(int posA, int posB) {
      Object aux = this.arr[posA];
      this.arr[posA] = this.arr[posB];
      this.arr[posB] = aux;
   }

   private void siftUp(int pos) {
      if (pos == 1) {
         return;
      }

      // comparo con el padre
      int parentPos = pos / 2;

      T elem = (T) this.arr[pos];
      T parent = (T) this.arr[parentPos];

      // parent > elem
      if (parent.compareTo(elem) > 0) {
         swap(pos, parentPos);
         siftUp(parentPos);
      }
   }

   public MinHeap(int expectedSize) {
      this.arr = new Object[expectedSize + 1];
      this.count = 0;
   }

   @Override
   public void push(T elem) {
      if (this.count == this.arr.length) {
         grow(this.arr.length * 2);
      }

      this.count++;
      this.arr[this.count] = elem;

      siftUp(this.count);
   }

   @Override
   public T top() {
      if (size() < 1)
         throw new RuntimeException("no hay elementos");

      return (T) this.arr[1];
   }

   @Override
   public T pop() {
      // TODO Auto-generated method stub
      throw new UnsupportedOperationException("Unimplemented method 'pop'");
   }

   @Override
   public int size() {
      return this.count;
   }

   @Override
   public boolean isEmpty() {
      return size() == 0;
   }

}
